commit c2fe38ca85272903c9c763f7f8ee1fec0818982e
Author: Bryan Drewery <bryan@shatow.net>
Date:   Sat Nov 19 16:39:09 2022 -0800

    sh: Track function for PS4

diff --git external/sh/expand.c external/sh/expand.c
index b08101a7d..c8ef3b187 100644
--- external/sh/expand.c
+++ external/sh/expand.c
@@ -953,6 +953,9 @@ varvalue(const char *name, int quoted, int subtype, int flag,
 	case '%':
 		num = shpid;
 		break;
+	case '+':
+		num = funcnest;
+		break;
 	case '?':
 		num = oexitstatus;
 		break;
diff --git external/sh/mksyntax.c external/sh/mksyntax.c
index fc92ff1a0..55346d691 100644
--- external/sh/mksyntax.c
+++ external/sh/mksyntax.c
@@ -225,7 +225,7 @@ main(int argc __unused, char **argv __unused)
 	add("abcdefghijklmnopqrstuvwxyz", "ISLOWER");
 	add("ABCDEFGHIJKLMNOPQRSTUVWXYZ", "ISUPPER");
 	add("_", "ISUNDER");
-	add("#%?$!-*@", "ISSPECL");
+	add("#%+?$!-*@", "ISSPECL");
 	finish();
 
 	exit(0);
