commit 04bedfe2d10a45b6e61b7454346799a484aed2bf
Author: Bryan Drewery <bryan@shatow.net>
Date:   Mon Dec 19 08:30:43 2022 -0800

    sh: Add $FUNCNAMESTACK

diff --git external/sh/eval.c external/sh/eval.c
index cd044c408..db9827166 100644
--- external/sh/eval.c
+++ external/sh/eval.c
@@ -930,6 +930,7 @@ safe_builtin(int idx, int argc, char **argv)
  * Note: This may or may not return if (flags & EV_EXIT).
  */
 
+int in_trap(void);
 static void
 evalcommand(union node *cmd, int flags, struct backcmd *backcmd)
 {
@@ -1166,6 +1167,25 @@ evalcommand(union node *cmd, int flags, struct backcmd *backcmd)
 		funcnest++;
 		mklocal("FUNCNAME");
 		setvar("FUNCNAME", argv[0], 0);
+		/* This can stay only if we ignore EPIPE. */
+		if (!in_trap()) {
+			char *funcstack;
+			int exitstatus_save = exitstatus, oexitstatus_save = oexitstatus;
+			mklocal("FUNCNAMESTACK");
+			asprintf(&funcstack,
+			    "FUNCNAMESTACK=\"${FUNCNAMESTACK:-${0}:%d:}${FUNCNAMESTACK:+:}${FUNCNAME}\"", plinno);
+			evalstring(funcstack, 0);
+			if (!is_int_on()) {
+				/*
+				 * evalstring may have FORCEINTON
+				 */
+				INTOFF;
+			}
+			assert(is_int_on());
+			exitstatus = exitstatus_save;
+			oexitstatus = oexitstatus_save;
+			free(funcstack);
+		}
 		redirect(cmd->ncmd.redirect, REDIR_PUSH);
 		INTON;
 		for (i = 0; i < varlist.count; i++)
diff --git external/sh/trap.c external/sh/trap.c
index 4d4b9ad01..679a92689 100644
--- external/sh/trap.c
+++ external/sh/trap.c
@@ -463,6 +463,7 @@ dotrap(void)
 	in_dotrap--;
 }
 
+int in_trap(void) { return in_dotrap; }
 
 void
 trap_init(void)
