diff --git external/sh/miscbltin.c external/sh/miscbltin.c
index bd4164811..d52433047 100644
--- external/sh/miscbltin.c
+++ external/sh/miscbltin.c
@@ -231,6 +231,8 @@ readcmd(int argc __unused, char **argv __unused)
 		 * If there's nothing ready, return an error.
 		 */
 		if (status <= 0) {
+			while (*ap != NULL)
+				setvar(*ap++, "", 0);
 			sig = pendingsig;
 			return (128 + (sig != 0 ? sig : SIGALRM));
 		}
