diff --git external/sh/error.c external/sh/error.c
index ade63b719..cace70ee1 100644
--- external/sh/error.c
+++ external/sh/error.c
@@ -130,9 +130,9 @@ void
 vwarning(const char *msg, va_list ap)
 {
 	if (commandname)
-		outfmt(out2, "%s: ", commandname);
+		outfmt(out2, "Error: (%d) %s: ", getpid(), commandname);
 	else if (arg0)
-		outfmt(out2, "%s: ", arg0);
+		outfmt(out2, "Error: (%d) %s: ", getpid(), arg0);
 	doformat(out2, msg, ap);
 	out2fmt_flush("\n");
 }
